[BUG] test_write_round_trip failed Maximum pool size exceeded

**Describe the bug**
We saw
```
FAILED ../../src/main/python/parquet_write_test.py::test_write_round_trip[TIMESTAMP_MILLIS-parquet-reader_confs2-[Byte, Short, Integer, Long, Float, Double, String, Boolean, Date, Timestamp, Struct(['child0', Byte],['child1', Short],['child2', Integer],['child3', Long],['child4', Float],['child5', Double],['child6', String],['child7', Boolean],['child8', Date],['child9', Timestamp]), Struct(['child0', Struct(['child1', Byte])]), Struct(['child0', Map(String(not_null),String)],['child1', Integer]), Array(Byte), Array(Short), Array(Integer), Array(Long), Array(Float), Array(Double), Array(String), Array(Boolean), Array(Date), Array(Timestamp), Array(Struct(['child0', Byte],['child1', Short],['child2', Integer],['child3', Long],['child4', Float],['child5', Double],['child6', String],['child7', Boolean],['child8', Date],['child9', Timestamp])), Array(Struct(['child0', Struct(['child1', Byte])])), Array(Struct(['child0', Map(String(not_null),String)],['child1', Integer])), Array(Array(Byte)), Array(Array(Short)), Array(Array(Integer)), Array(Array(Long)), Array(Array(Float)), Array(Array(Double)), Array(Array(String)), Array(Array(Boolean)), Array(Array(Date)), Array(Array(Timestamp)), Array(Array(Struct(['child0', Byte],['child1', Short],['child2', Integer],['child3', Long],['child4', Float],['child5', Double],['child6', String],['child7', Boolean],['child8', Date],['child9', Timestamp]))), Array(Array(Struct(['child0', Struct(['child1', Byte])]))), Array(Array(Struct(['child0', Map(String(not_null),String)],['child1', Integer]))), Decimal(18,0), Decimal(7,3), Decimal(7,7), Decimal(12,2), Decimal(20,2), Decimal(36,5), Decimal(38,0), Decimal(38,10), Map(Boolean(not_null),Boolean), Map(Byte(not_null),Byte), Map(Short(not_null),Short), Map(Integer(not_null),Integer), Map(Long(not_null),Long), Map(Float(not_null),Float), Map(Double(not_null),Double), Map(Date(not_null),Date), Map(Timestamp(not_null),Timestamp), Map(String(not_null),String), Map(not_null)(Decimal(not_null)(20,2),Decimal(not_null)(20,2)), Map(not_null)(Decimal(not_null)(36,5),Decimal(not_null)(36,5)), Map(not_null)(Decimal(not_null)(38,38),Decimal(not_null)(38,38)), Map(String(not_null),Array(String)), Map(RepeatSeq(not_null)(Integer(not_null)),Long), Map(String(not_null),Map(String(not_null),String)), Map(Struct(not_null)(['child0', String],['child1', String]),Float), Map(Struct(not_null)(['child0', String]),String)]]
```
in nightly integration test (spark312 only). rapids_integration-dev-github build ID 375

```
[2022-01-15T21:25:04.922Z]     def get_return_value(answer, gateway_client, target_id=None, name=None):

[2022-01-15T21:25:04.922Z]         """Converts an answer received from the Java gateway into a Python object.

[2022-01-15T21:25:04.922Z]     

[2022-01-15T21:25:04.922Z]         For example, string representation of integers are converted to Python

[2022-01-15T21:25:04.922Z]         integer, string representation of objects are converted to JavaObject

[2022-01-15T21:25:04.922Z]         instances, etc.

[2022-01-15T21:25:04.922Z]     

[2022-01-15T21:25:04.922Z]         :param answer: the string returned by the Java gateway

[2022-01-15T21:25:04.922Z]         :param gateway_client: the gateway client used to communicate with the Java

[2022-01-15T21:25:04.922Z]             Gateway. Only necessary if the answer is a reference (e.g., object,

[2022-01-15T21:25:04.922Z]             list, map)

[2022-01-15T21:25:04.922Z]         :param target_id: the name of the object from which the answer comes from

[2022-01-15T21:25:04.922Z]             (e.g., *object1* in `object1.hello()`). Optional.

[2022-01-15T21:25:04.922Z]         :param name: the name of the member from which the answer comes from

[2022-01-15T21:25:04.922Z]             (e.g., *hello* in `object1.hello()`). Optional.

[2022-01-15T21:25:04.922Z]         """

[2022-01-15T21:25:04.922Z]         if is_error(answer)[0]:

[2022-01-15T21:25:04.922Z]             if len(answer) > 1:

[2022-01-15T21:25:04.922Z]                 type = answer[1]

[2022-01-15T21:25:04.922Z]                 value = OUTPUT_CONVERTER[type](answer[2:], gateway_client)

[2022-01-15T21:25:04.922Z]                 if answer[1] == REFERENCE_TYPE:

[2022-01-15T21:25:04.922Z] >                   raise Py4JJavaError(

[2022-01-15T21:25:04.922Z]                         "An error occurred while calling {0}{1}{2}.\n".

[2022-01-15T21:25:04.922Z]                         format(target_id, ".", name), value)

[2022-01-15T21:25:04.922Z] E                   py4j.protocol.Py4JJavaError: An error occurred while calling o8415.parquet.

[2022-01-15T21:25:04.922Z] E                   : org.apache.spark.SparkException: Job aborted.

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.write(GpuFileFormatWriter.scala:259)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.rapids.GpuInsertIntoHadoopFsRelationCommand.runColumnar(GpuInsertIntoHadoopFsRelationCommand.scala:166)

[2022-01-15T21:25:04.922Z] E                   	at com.nvidia.spark.rapids.GpuDataWritingCommandExec.sideEffectResult$lzycompute(GpuDataWritingCommandExec.scala:97)

[2022-01-15T21:25:04.922Z] E                   	at com.nvidia.spark.rapids.GpuDataWritingCommandExec.sideEffectResult(GpuDataWritingCommandExec.scala:96)

[2022-01-15T21:25:04.922Z] E                   	at com.nvidia.spark.rapids.GpuDataWritingCommandExec.doExecuteColumnar(GpuDataWritingCommandExec.scala:120)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:207)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:203)

[2022-01-15T21:25:04.922Z] E                   	at com.nvidia.spark.rapids.GpuColumnarToRowExecParent.doExecute(GpuColumnarToRowExec.scala:319)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:180)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:218)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:215)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:176)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.QueryExecution.toRdd$lzycompute(QueryExecution.scala:132)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.QueryExecution.toRdd(QueryExecution.scala:131)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.DataFrameWriter.$anonfun$runCommand$1(DataFrameWriter.scala:989)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:775)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.DataFrameWriter.runCommand(DataFrameWriter.scala:989)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.DataFrameWriter.saveToV1Source(DataFrameWriter.scala:438)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.DataFrameWriter.saveInternal(DataFrameWriter.scala:415)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.DataFrameWriter.save(DataFrameWriter.scala:293)

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.DataFrameWriter.parquet(DataFrameWriter.scala:874)

[2022-01-15T21:25:04.922Z] E                   	at sun.reflect.GeneratedMethodAccessor114.invoke(Unknown Source)

[2022-01-15T21:25:04.922Z] E                   	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

[2022-01-15T21:25:04.922Z] E                   	at java.lang.reflect.Method.invoke(Method.java:498)

[2022-01-15T21:25:04.922Z] E                   	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)

[2022-01-15T21:25:04.922Z] E                   	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)

[2022-01-15T21:25:04.922Z] E                   	at py4j.Gateway.invoke(Gateway.java:282)

[2022-01-15T21:25:04.922Z] E                   	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)

[2022-01-15T21:25:04.922Z] E                   	at py4j.commands.CallCommand.execute(CallCommand.java:79)

[2022-01-15T21:25:04.922Z] E                   	at py4j.GatewayConnection.run(GatewayConnection.java:238)

[2022-01-15T21:25:04.922Z] E                   	at java.lang.Thread.run(Thread.java:748)

[2022-01-15T21:25:04.922Z] E                   Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 151.0 failed 1 times, most recent failure: Lost task 0.0 in stage 151.0 (TID 185) (10.233.92.241 executor 0): org.apache.spark.SparkException: Task failed while writing rows.

[2022-01-15T21:25:04.922Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:320)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$write$14(GpuFileFormatWriter.scala:238)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.Task.run(Task.scala:131)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)

[2022-01-15T21:25:04.923Z] E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

[2022-01-15T21:25:04.923Z] E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

[2022-01-15T21:25:04.923Z] E                   	at java.lang.Thread.run(Thread.java:748)

[2022-01-15T21:25:04.923Z] E                   Caused by: java.lang.OutOfMemoryError: Could not allocate native memory: std::bad_alloc: out_of_memory: RMM failure at:/home/jenkins/agent/workspace/jenkins-cudf_nightly-dev-github-599-cuda11/cpp/build/_deps/rmm-src/include/rmm/mr/device/arena_memory_resource.hpp:159: Maximum pool size exceeded

[2022-01-15T21:25:04.923Z] E                   	at ai.rapids.cudf.Table.writeParquetChunk(Native Method)

[2022-01-15T21:25:04.923Z] E                   	at ai.rapids.cudf.Table.access$300(Table.java:41)

[2022-01-15T21:25:04.923Z] E                   	at ai.rapids.cudf.Table$ParquetTableWriter.write(Table.java:1119)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2(ColumnarOutputWriter.scala:137)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2$adapted(ColumnarOutputWriter.scala:134)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1(ColumnarOutputWriter.scala:134)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1$adapted(ColumnarOutputWriter.scala:133)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.writeBatch(ColumnarOutputWriter.scala:133)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.write(ColumnarOutputWriter.scala:99)

[2022-01-15T21:25:04.923Z] E                   	at com.nvidia.spark.rapids.GpuParquetWriter.write(GpuParquetFileFormat.scala:339)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.sql.rapids.GpuSingleDirectoryDataWriter.write(GpuFileFormatDataWriter.scala:163)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$executeTask$1(GpuFileFormatWriter.scala:306)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1473)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:314)

[2022-01-15T21:25:04.923Z] E                   	... 9 more

[2022-01-15T21:25:04.923Z] E                   

[2022-01-15T21:25:04.923Z] E                   Driver stacktrace:

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2258)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2207)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2206)

[2022-01-15T21:25:04.923Z] E                   	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)

[2022-01-15T21:25:04.923Z] E                   	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)

[2022-01-15T21:25:04.923Z] E                   	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2206)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1079)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1079)

[2022-01-15T21:25:04.923Z] E                   	at scala.Option.foreach(Option.scala:407)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1079)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2445)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2387)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2376)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:868)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2196)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.write(GpuFileFormatWriter.scala:228)

[2022-01-15T21:25:04.923Z] E                   	... 38 more

[2022-01-15T21:25:04.923Z] E                   Caused by: org.apache.spark.SparkException: Task failed while writing rows.

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:320)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$write$14(GpuFileFormatWriter.scala:238)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.scheduler.Task.run(Task.scala:131)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)

[2022-01-15T21:25:04.923Z] E                   	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)

[2022-01-15T21:25:04.924Z] E                   	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)

[2022-01-15T21:25:04.924Z] E                   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

[2022-01-15T21:25:04.924Z] E                   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

[2022-01-15T21:25:04.924Z] E                   	... 1 more

[2022-01-15T21:25:04.924Z] E                   Caused by: java.lang.OutOfMemoryError: Could not allocate native memory: std::bad_alloc: out_of_memory: RMM failure at:/home/jenkins/agent/workspace/jenkins-cudf_nightly-dev-github-599-cuda11/cpp/build/_deps/rmm-src/include/rmm/mr/device/arena_memory_resource.hpp:159: Maximum pool size exceeded

[2022-01-15T21:25:04.924Z] E                   	at ai.rapids.cudf.Table.writeParquetChunk(Native Method)

[2022-01-15T21:25:04.924Z] E                   	at ai.rapids.cudf.Table.access$300(Table.java:41)

[2022-01-15T21:25:04.924Z] E                   	at ai.rapids.cudf.Table$ParquetTableWriter.write(Table.java:1119)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2(ColumnarOutputWriter.scala:137)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2$adapted(ColumnarOutputWriter.scala:134)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1(ColumnarOutputWriter.scala:134)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1$adapted(ColumnarOutputWriter.scala:133)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.writeBatch(ColumnarOutputWriter.scala:133)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.ColumnarOutputWriter.write(ColumnarOutputWriter.scala:99)

[2022-01-15T21:25:04.924Z] E                   	at com.nvidia.spark.rapids.GpuParquetWriter.write(GpuParquetFileFormat.scala:339)

[2022-01-15T21:25:04.924Z] E                   	at org.apache.spark.sql.rapids.GpuSingleDirectoryDataWriter.write(GpuFileFormatDataWriter.scala:163)

[2022-01-15T21:25:04.924Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$executeTask$1(GpuFileFormatWriter.scala:306)

[2022-01-15T21:25:04.924Z] E                   	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1473)

[2022-01-15T21:25:04.924Z] E                   	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:314)

[2022-01-15T21:25:04.924Z] E                   	... 9 more
```

This only happened once in spark312 nightly integration test, since this is the first time we saw it.
Please help confirm if
the test does require more pool size comparing to before (parallel run use smaller rmm pool size for each test)
or 
there could be some memory leak somewhere

add executor log

```
22/01/15 21:24:59 INFO ParquetWriteSupport: Initialized Parquet WriteSupport with Catalyst schema:
{
  "type" : "struct",
  "fields" : [ {
    "name" : "_c0",
    "type" : "byte",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c1",
    "type" : "short",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c2",
    "type" : "integer",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c3",
    "type" : "long",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c4",
    "type" : "float",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c5",
    "type" : "double",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c6",
    "type" : "string",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c7",
    "type" : "boolean",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c8",
    "type" : "date",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c9",
    "type" : "timestamp",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c10",
    "type" : {
      "type" : "struct",
      "fields" : [ {
        "name" : "child0",
        "type" : "byte",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child1",
        "type" : "short",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child2",
        "type" : "integer",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child3",
        "type" : "long",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child4",
        "type" : "float",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child5",
        "type" : "double",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child6",
        "type" : "string",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child7",
        "type" : "boolean",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child8",
        "type" : "date",
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child9",
        "type" : "timestamp",
        "nullable" : true,
        "metadata" : { }
      } ]
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c11",
    "type" : {
      "type" : "struct",
      "fields" : [ {
        "name" : "child0",
        "type" : {
          "type" : "struct",
          "fields" : [ {
            "name" : "child1",
            "type" : "byte",
            "nullable" : true,
            "metadata" : { }
          } ]
        },
        "nullable" : true,
        "metadata" : { }
      } ]
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c12",
    "type" : {
      "type" : "struct",
      "fields" : [ {
        "name" : "child0",
        "type" : {
          "type" : "map",
          "keyType" : "string",
          "valueType" : "string",
          "valueContainsNull" : true
        },
        "nullable" : true,
        "metadata" : { }
      }, {
        "name" : "child1",
        "type" : "integer",
        "nullable" : true,
        "metadata" : { }
      } ]
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c13",
    "type" : {
      "type" : "array",
      "elementType" : "byte",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c14",
    "type" : {
      "type" : "array",
      "elementType" : "short",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c15",
    "type" : {
      "type" : "array",
      "elementType" : "integer",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c16",
    "type" : {
      "type" : "array",
      "elementType" : "long",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c17",
    "type" : {
      "type" : "array",
      "elementType" : "float",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c18",
    "type" : {
      "type" : "array",
      "elementType" : "double",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c19",
    "type" : {
      "type" : "array",
      "elementType" : "string",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c20",
    "type" : {
      "type" : "array",
      "elementType" : "boolean",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c21",
    "type" : {
      "type" : "array",
      "elementType" : "date",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c22",
    "type" : {
      "type" : "array",
      "elementType" : "timestamp",
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c23",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "struct",
        "fields" : [ {
          "name" : "child0",
          "type" : "byte",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child1",
          "type" : "short",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child2",
          "type" : "integer",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child3",
          "type" : "long",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child4",
          "type" : "float",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child5",
          "type" : "double",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child6",
          "type" : "string",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child7",
          "type" : "boolean",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child8",
          "type" : "date",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child9",
          "type" : "timestamp",
          "nullable" : true,
          "metadata" : { }
        } ]
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c24",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "struct",
        "fields" : [ {
          "name" : "child0",
          "type" : {
            "type" : "struct",
            "fields" : [ {
              "name" : "child1",
              "type" : "byte",
              "nullable" : true,
              "metadata" : { }
            } ]
          },
          "nullable" : true,
          "metadata" : { }
        } ]
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c25",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "struct",
        "fields" : [ {
          "name" : "child0",
          "type" : {
            "type" : "map",
            "keyType" : "string",
            "valueType" : "string",
            "valueContainsNull" : true
          },
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child1",
          "type" : "integer",
          "nullable" : true,
          "metadata" : { }
        } ]
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c26",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "byte",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c27",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "short",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c28",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "integer",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c29",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "long",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c30",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "float",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c31",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "double",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c32",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "string",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c33",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "boolean",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c34",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "date",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c35",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : "timestamp",
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c36",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : {
          "type" : "struct",
          "fields" : [ {
            "name" : "child0",
            "type" : "byte",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child1",
            "type" : "short",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child2",
            "type" : "integer",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child3",
            "type" : "long",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child4",
            "type" : "float",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child5",
            "type" : "double",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child6",
            "type" : "string",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child7",
            "type" : "boolean",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child8",
            "type" : "date",
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child9",
            "type" : "timestamp",
            "nullable" : true,
            "metadata" : { }
          } ]
        },
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c37",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : {
          "type" : "struct",
          "fields" : [ {
            "name" : "child0",
            "type" : {
              "type" : "struct",
              "fields" : [ {
                "name" : "child1",
                "type" : "byte",
                "nullable" : true,
                "metadata" : { }
              } ]
            },
            "nullable" : true,
            "metadata" : { }
          } ]
        },
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c38",
    "type" : {
      "type" : "array",
      "elementType" : {
        "type" : "array",
        "elementType" : {
          "type" : "struct",
          "fields" : [ {
            "name" : "child0",
            "type" : {
              "type" : "map",
              "keyType" : "string",
              "valueType" : "string",
              "valueContainsNull" : true
            },
            "nullable" : true,
            "metadata" : { }
          }, {
            "name" : "child1",
            "type" : "integer",
            "nullable" : true,
            "metadata" : { }
          } ]
        },
        "containsNull" : true
      },
      "containsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c39",
    "type" : "decimal(18,0)",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c40",
    "type" : "decimal(7,3)",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c41",
    "type" : "decimal(7,7)",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c42",
    "type" : "decimal(12,2)",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c43",
    "type" : "decimal(20,2)",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c44",
    "type" : "decimal(36,5)",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c45",
    "type" : "decimal(38,0)",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c46",
    "type" : "decimal(38,10)",
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c47",
    "type" : {
      "type" : "map",
      "keyType" : "boolean",
      "valueType" : "boolean",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c48",
    "type" : {
      "type" : "map",
      "keyType" : "byte",
      "valueType" : "byte",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c49",
    "type" : {
      "type" : "map",
      "keyType" : "short",
      "valueType" : "short",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c50",
    "type" : {
      "type" : "map",
      "keyType" : "integer",
      "valueType" : "integer",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c51",
    "type" : {
      "type" : "map",
      "keyType" : "long",
      "valueType" : "long",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c52",
    "type" : {
      "type" : "map",
      "keyType" : "float",
      "valueType" : "float",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c53",
    "type" : {
      "type" : "map",
      "keyType" : "double",
      "valueType" : "double",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c54",
    "type" : {
      "type" : "map",
      "keyType" : "date",
      "valueType" : "date",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c55",
    "type" : {
      "type" : "map",
      "keyType" : "timestamp",
      "valueType" : "timestamp",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c56",
    "type" : {
      "type" : "map",
      "keyType" : "string",
      "valueType" : "string",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c57",
    "type" : {
      "type" : "map",
      "keyType" : "decimal(20,2)",
      "valueType" : "decimal(20,2)",
      "valueContainsNull" : false
    },
    "nullable" : false,
    "metadata" : { }
  }, {
    "name" : "_c58",
    "type" : {
      "type" : "map",
      "keyType" : "decimal(36,5)",
      "valueType" : "decimal(36,5)",
      "valueContainsNull" : false
    },
    "nullable" : false,
    "metadata" : { }
  }, {
    "name" : "_c59",
    "type" : {
      "type" : "map",
      "keyType" : "decimal(38,38)",
      "valueType" : "decimal(38,38)",
      "valueContainsNull" : false
    },
    "nullable" : false,
    "metadata" : { }
  }, {
    "name" : "_c60",
    "type" : {
      "type" : "map",
      "keyType" : "string",
      "valueType" : {
        "type" : "array",
        "elementType" : "string",
        "containsNull" : true
      },
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c61",
    "type" : {
      "type" : "map",
      "keyType" : "integer",
      "valueType" : "long",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c62",
    "type" : {
      "type" : "map",
      "keyType" : "string",
      "valueType" : {
        "type" : "map",
        "keyType" : "string",
        "valueType" : "string",
        "valueContainsNull" : true
      },
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c63",
    "type" : {
      "type" : "map",
      "keyType" : {
        "type" : "struct",
        "fields" : [ {
          "name" : "child0",
          "type" : "string",
          "nullable" : true,
          "metadata" : { }
        }, {
          "name" : "child1",
          "type" : "string",
          "nullable" : true,
          "metadata" : { }
        } ]
      },
      "valueType" : "float",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  }, {
    "name" : "_c64",
    "type" : {
      "type" : "map",
      "keyType" : {
        "type" : "struct",
        "fields" : [ {
          "name" : "child0",
          "type" : "string",
          "nullable" : true,
          "metadata" : { }
        } ]
      },
      "valueType" : "string",
      "valueContainsNull" : true
    },
    "nullable" : true,
    "metadata" : { }
  } ]
}
and corresponding Parquet message type:
message spark_schema {
  optional int32 _c0 (INT_8);
  optional int32 _c1 (INT_16);
  optional int32 _c2;
  optional int64 _c3;
  optional float _c4;
  optional double _c5;
  optional binary _c6 (UTF8);
  optional boolean _c7;
  optional int32 _c8 (DATE);
  optional int64 _c9 (TIMESTAMP_MILLIS);
  optional group _c10 {
    optional int32 child0 (INT_8);
    optional int32 child1 (INT_16);
    optional int32 child2;
    optional int64 child3;
    optional float child4;
    optional double child5;
    optional binary child6 (UTF8);
    optional boolean child7;
    optional int32 child8 (DATE);
    optional int64 child9 (TIMESTAMP_MILLIS);
  }
  optional group _c11 {
    optional group child0 {
      optional int32 child1 (INT_8);
    }
  }
  optional group _c12 {
    optional group child0 (MAP) {
      repeated group key_value {
        required binary key (UTF8);
        optional binary value (UTF8);
      }
    }
    optional int32 child1;
  }
  optional group _c13 (LIST) {
    repeated group list {
      optional int32 element (INT_8);
    }
  }
  optional group _c14 (LIST) {
    repeated group list {
      optional int32 element (INT_16);
    }
  }
  optional group _c15 (LIST) {
    repeated group list {
      optional int32 element;
    }
  }
  optional group _c16 (LIST) {
    repeated group list {
      optional int64 element;
    }
  }
  optional group _c17 (LIST) {
    repeated group list {
      optional float element;
    }
  }
  optional group _c18 (LIST) {
    repeated group list {
      optional double element;
    }
  }
  optional group _c19 (LIST) {
    repeated group list {
      optional binary element (UTF8);
    }
  }
  optional group _c20 (LIST) {
    repeated group list {
      optional boolean element;
    }
  }
  optional group _c21 (LIST) {
    repeated group list {
      optional int32 element (DATE);
    }
  }
  optional group _c22 (LIST) {
    repeated group list {
      optional int64 element (TIMESTAMP_MILLIS);
    }
  }
  optional group _c23 (LIST) {
    repeated group list {
      optional group element {
        optional int32 child0 (INT_8);
        optional int32 child1 (INT_16);
        optional int32 child2;
        optional int64 child3;
        optional float child4;
        optional double child5;
        optional binary child6 (UTF8);
        optional boolean child7;
        optional int32 child8 (DATE);
        optional int64 child9 (TIMESTAMP_MILLIS);
      }
    }
  }
  optional group _c24 (LIST) {
    repeated group list {
      optional group element {
        optional group child0 {
          optional int32 child1 (INT_8);
        }
      }
    }
  }
  optional group _c25 (LIST) {
    repeated group list {
      optional group element {
        optional group child0 (MAP) {
          repeated group key_value {
            required binary key (UTF8);
            optional binary value (UTF8);
          }
        }
        optional int32 child1;
      }
    }
  }
  optional group _c26 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional int32 element (INT_8);
        }
      }
    }
  }
  optional group _c27 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional int32 element (INT_16);
        }
      }
    }
  }
  optional group _c28 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional int32 element;
        }
      }
    }
  }
  optional group _c29 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional int64 element;
        }
      }
    }
  }
  optional group _c30 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional float element;
        }
      }
    }
  }
  optional group _c31 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional double element;
        }
      }
    }
  }
  optional group _c32 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional binary element (UTF8);
        }
      }
    }
  }
  optional group _c33 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional boolean element;
        }
      }
    }
  }
  optional group _c34 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional int32 element (DATE);
        }
      }
    }
  }
  optional group _c35 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional int64 element (TIMESTAMP_MILLIS);
        }
      }
    }
  }
  optional group _c36 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional group element {
            optional int32 child0 (INT_8);
            optional int32 child1 (INT_16);
            optional int32 child2;
            optional int64 child3;
            optional float child4;
            optional double child5;
            optional binary child6 (UTF8);
            optional boolean child7;
            optional int32 child8 (DATE);
            optional int64 child9 (TIMESTAMP_MILLIS);
          }
        }
      }
    }
  }
  optional group _c37 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional group element {
            optional group child0 {
              optional int32 child1 (INT_8);
            }
          }
        }
      }
    }
  }
  optional group _c38 (LIST) {
    repeated group list {
      optional group element (LIST) {
        repeated group list {
          optional group element {
            optional group child0 (MAP) {
              repeated group key_value {
                required binary key (UTF8);
                optional binary value (UTF8);
              }
            }
            optional int32 child1;
          }
        }
      }
    }
  }
  optional int64 _c39 (DECIMAL(18,0));
  optional int32 _c40 (DECIMAL(7,3));
  optional int32 _c41 (DECIMAL(7,7));
  optional int64 _c42 (DECIMAL(12,2));
  optional fixed_len_byte_array(9) _c43 (DECIMAL(20,2));
  optional fixed_len_byte_array(16) _c44 (DECIMAL(36,5));
  optional fixed_len_byte_array(16) _c45 (DECIMAL(38,0));
  optional fixed_len_byte_array(16) _c46 (DECIMAL(38,10));
  optional group _c47 (MAP) {
    repeated group key_value {
      required boolean key;
      optional boolean value;
    }
  }
  optional group _c48 (MAP) {
    repeated group key_value {
      required int32 key (INT_8);
      optional int32 value (INT_8);
    }
  }
  optional group _c49 (MAP) {
    repeated group key_value {
      required int32 key (INT_16);
      optional int32 value (INT_16);
    }
  }
  optional group _c50 (MAP) {
    repeated group key_value {
      required int32 key;
      optional int32 value;
    }
  }
  optional group _c51 (MAP) {
    repeated group key_value {
      required int64 key;
      optional int64 value;
    }
  }
  optional group _c52 (MAP) {
    repeated group key_value {
      required float key;
      optional float value;
    }
  }
  optional group _c53 (MAP) {
    repeated group key_value {
      required double key;
      optional double value;
    }
  }
  optional group _c54 (MAP) {
    repeated group key_value {
      required int32 key (DATE);
      optional int32 value (DATE);
    }
  }
  optional group _c55 (MAP) {
    repeated group key_value {
      required int64 key (TIMESTAMP_MILLIS);
      optional int64 value (TIMESTAMP_MILLIS);
    }
  }
  optional group _c56 (MAP) {
    repeated group key_value {
      required binary key (UTF8);
      optional binary value (UTF8);
    }
  }
  required group _c57 (MAP) {
    repeated group key_value {
      required fixed_len_byte_array(9) key (DECIMAL(20,2));
      required fixed_len_byte_array(9) value (DECIMAL(20,2));
    }
  }
  required group _c58 (MAP) {
    repeated group key_value {
      required fixed_len_byte_array(16) key (DECIMAL(36,5));
      required fixed_len_byte_array(16) value (DECIMAL(36,5));
    }
  }
  required group _c59 (MAP) {
    repeated group key_value {
      required fixed_len_byte_array(16) key (DECIMAL(38,38));
      required fixed_len_byte_array(16) value (DECIMAL(38,38));
    }
  }
  optional group _c60 (MAP) {
    repeated group key_value {
      required binary key (UTF8);
      optional group value (LIST) {
        repeated group list {
          optional binary element (UTF8);
        }
      }
    }
  }
  optional group _c61 (MAP) {
    repeated group key_value {
      required int32 key;
      optional int64 value;
    }
  }
  optional group _c62 (MAP) {
    repeated group key_value {
      required binary key (UTF8);
      optional group value (MAP) {
        repeated group key_value {
          required binary key (UTF8);
          optional binary value (UTF8);
        }
      }
    }
  }
  optional group _c63 (MAP) {
    repeated group key_value {
      required group key {
        optional binary child0 (UTF8);
        optional binary child1 (UTF8);
      }
      optional float value;
    }
  }
  optional group _c64 (MAP) {
    repeated group key_value {
      required group key {
        optional binary child0 (UTF8);
      }
      optional binary value (UTF8);
    }
  }
}

       
22/01/15 21:25:02 INFO PythonRunner: Times: total = 2646, boot = 3, init = 15, finish = 2628
22/01/15 21:25:02 INFO DeviceMemoryEventHandler: Device allocation of 578025832 bytes failed, device store has 0 bytes. Total RMM allocated is 37888256 bytes.
22/01/15 21:25:02 WARN DeviceMemoryEventHandler: Device store exhausted, unable to allocate 578025832 bytes. Total RMM allocated is 37888256 bytes.
22/01/15 21:25:02 ERROR Utils: Aborting task
java.lang.OutOfMemoryError: Could not allocate native memory: std::bad_alloc: out_of_memory: RMM failure at:/home/jenkins/agent/workspace/jenkins-cudf_nightly-dev-github-599-cuda11/cpp/build/_deps/rmm-src/include/rmm/mr/device/arena_memory_resource.hpp:159: Maximum pool size exceeded
	at ai.rapids.cudf.Table.writeParquetChunk(Native Method)
	at ai.rapids.cudf.Table.access$300(Table.java:41)
	at ai.rapids.cudf.Table$ParquetTableWriter.write(Table.java:1119)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2(ColumnarOutputWriter.scala:137)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2$adapted(ColumnarOutputWriter.scala:134)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1(ColumnarOutputWriter.scala:134)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1$adapted(ColumnarOutputWriter.scala:133)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.writeBatch(ColumnarOutputWriter.scala:133)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.write(ColumnarOutputWriter.scala:99)
	at com.nvidia.spark.rapids.GpuParquetWriter.write(GpuParquetFileFormat.scala:339)
	at org.apache.spark.sql.rapids.GpuSingleDirectoryDataWriter.write(GpuFileFormatDataWriter.scala:163)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$executeTask$1(GpuFileFormatWriter.scala:306)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1473)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:314)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$write$14(GpuFileFormatWriter.scala:238)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
22/01/15 21:25:02 ERROR GpuFileFormatWriter: Job job_202201152124597348919861576254515_0151 aborted.
22/01/15 21:25:02 ERROR Executor: Exception in task 0.0 in stage 151.0 (TID 185)
org.apache.spark.SparkException: Task failed while writing rows.
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:320)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$write$14(GpuFileFormatWriter.scala:238)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.OutOfMemoryError: Could not allocate native memory: std::bad_alloc: out_of_memory: RMM failure at:/home/jenkins/agent/workspace/jenkins-cudf_nightly-dev-github-599-cuda11/cpp/build/_deps/rmm-src/include/rmm/mr/device/arena_memory_resource.hpp:159: Maximum pool size exceeded
	at ai.rapids.cudf.Table.writeParquetChunk(Native Method)
	at ai.rapids.cudf.Table.access$300(Table.java:41)
	at ai.rapids.cudf.Table$ParquetTableWriter.write(Table.java:1119)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2(ColumnarOutputWriter.scala:137)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2$adapted(ColumnarOutputWriter.scala:134)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1(ColumnarOutputWriter.scala:134)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1$adapted(ColumnarOutputWriter.scala:133)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.writeBatch(ColumnarOutputWriter.scala:133)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.write(ColumnarOutputWriter.scala:99)
	at com.nvidia.spark.rapids.GpuParquetWriter.write(GpuParquetFileFormat.scala:339)
	at org.apache.spark.sql.rapids.GpuSingleDirectoryDataWriter.write(GpuFileFormatDataWriter.scala:163)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$executeTask$1(GpuFileFormatWriter.scala:306)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1473)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:314)
	... 9 more
22/01/15 21:25:02 ERROR SparkUncaughtExceptionHandler: Uncaught exception in thread Thread[Executor task launch worker for task 0.0 in stage 151.0 (TID 185),5,main]
org.apache.spark.SparkException: Task failed while writing rows.
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:320)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$write$14(GpuFileFormatWriter.scala:238)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.OutOfMemoryError: Could not allocate native memory: std::bad_alloc: out_of_memory: RMM failure at:/home/jenkins/agent/workspace/jenkins-cudf_nightly-dev-github-599-cuda11/cpp/build/_deps/rmm-src/include/rmm/mr/device/arena_memory_resource.hpp:159: Maximum pool size exceeded
	at ai.rapids.cudf.Table.writeParquetChunk(Native Method)
	at ai.rapids.cudf.Table.access$300(Table.java:41)
	at ai.rapids.cudf.Table$ParquetTableWriter.write(Table.java:1119)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2(ColumnarOutputWriter.scala:137)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$2$adapted(ColumnarOutputWriter.scala:134)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1(ColumnarOutputWriter.scala:134)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatch$1$adapted(ColumnarOutputWriter.scala:133)
	at com.nvidia.spark.rapids.Arm.withResource(Arm.scala:28)
	at com.nvidia.spark.rapids.Arm.withResource$(Arm.scala:26)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.withResource(ColumnarOutputWriter.scala:64)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.writeBatch(ColumnarOutputWriter.scala:133)
	at com.nvidia.spark.rapids.ColumnarOutputWriter.write(ColumnarOutputWriter.scala:99)
	at com.nvidia.spark.rapids.GpuParquetWriter.write(GpuParquetFileFormat.scala:339)
	at org.apache.spark.sql.rapids.GpuSingleDirectoryDataWriter.write(GpuFileFormatDataWriter.scala:163)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$executeTask$1(GpuFileFormatWriter.scala:306)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1473)
	at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:314)
	... 9 more
22/01/15 21:25:02 INFO RapidsBufferCatalog: Closing storage
22/01/15 21:25:02 INFO MemoryStore: MemoryStore cleared
22/01/15 21:25:02 INFO BlockManager: BlockManager stopped
```

--- a/integration_tests/run_pyspark_from_build.sh
+++ b/integration_tests/run_pyspark_from_build.sh
@@ -64,24 +64,32 @@ else
     if [[ "${TEST_PARALLEL}" == "" ]];
     then
         # For integration tests we want to have at least
-        #  - 2 GiB of GPU memory
-        #  - 5 GiB of host memory
+        #  - 1.5 GiB of GPU memory for the tests and 750 MiB for loading CUDF + CUDA.
+        #    From profiling we saw that tests allocated under 200 MiB of GPU memory and
+        #    1.5 GiB felt like it gave us plenty of room to grow.
+        #  - 5 GiB of host memory. In testing with a limited number of tasks (4) we saw
+        #    the amount of host memory not go above 3 GiB so 5 felt like a good number
+        #    for future growth.
         #  - 1 CPU core
-        # per Spark application. And we want to reserve 2 GiB of GPU memory for CUDA/etc.
+        # per Spark application. We reserve 2 GiB of GPU memory for general overhead also.
 
-        # For now just assume that we are going to use the GPU on the
-        # system with the most free memory.
-        # We use free memory to try and avoid issues if the GPU also is working
-        # on graphics, which happens some times.
-        # we also reserve 2 GiB for CUDA/CUDF overhead on the GPU.
-        
-        GPU_MEM_PARALLEL=`nvidia-smi --query-gpu=memory.free --format=csv,noheader | awk '{if (MAX < $1){ MAX = $1}} END {print int((MAX - 2 * 1024)/ ((1.5 * 1024) + 750))}'`
+        # For now just assume that we are going to use the GPU on the system with the most
+        # free memory. We use free memory to try and avoid issues if the GPU also is working
+        # on graphics, which happens with many workstation GPUs. We also reserve 2 GiB for
+        # CUDA/CUDF overhead which can be used because of JIT or launching large kernels.
+       
+        # If you need to increase the amount of GPU memory you need to change it here and
+        # below where the processes are launched.
+        GPU_MEM_PARALLEL=`nvidia-smi --query-gpu=memory.free --format=csv,noheader | awk '{if (MAX < $1){ MAX = $1}} END {print int((MAX - 2 * 1024) / ((1.5 * 1024) + 750))}'`
         CPU_CORES=`nproc` 
         HOST_MEM_PARALLEL=`cat /proc/meminfo | grep MemAvailable | awk '{print int($2 / (5 * 1024))}'`
-        TMP_PARALLEL=$(( $GPU_MEM_PARALLEL>$CPU_CORES ? $CPU_CORES : $GPU_MEM_PARALLEL ))
-        TMP_PARALLEL=$(( $TMP_PARALLEL>$HOST_MEM_PARALLEL ? $HOST_MEM_PARALLEL : $TMP_PARALLEL ))
+        TMP_PARALLEL=$(( $GPU_MEM_PARALLEL > $CPU_CORES ? $CPU_CORES : $GPU_MEM_PARALLEL ))
+        TMP_PARALLEL=$(( $TMP_PARALLEL > $HOST_MEM_PARALLEL ? $HOST_MEM_PARALLEL : $TMP_PARALLEL ))
         if [[ $TMP_PARALLEL -gt 1 ]];
         then
+            # We subtract 1 from the parallel number because xdist launches a process to
+            # control and monitor the other processes. It takes up one available parallel
+            # slot, even if it is not truly using all of the resources we give it.
             TEST_PARALLEL=$(( $TMP_PARALLEL - 1 ))
         else
             TEST_PARALLEL=1
@@ -160,7 +168,8 @@ else
     # prevent cluster shape to change
     export PYSP_TEST_spark_dynamicAllocation_enabled='false'
     export PYSP_TEST_spark_rapids_memory_host_spillStorageSize='100m'
-    # Not the default 2G but should be large enough for a single batch for all data
+    # Not the default 2G but should be large enough for a single batch for all data (we found
+    # 200 MiB being allocated by a single test at most, and we typically have 4 tasks.
     export PYSP_TEST_spark_rapids_sql_batchSizeBytes='100m'
 
     # Extract Databricks version from deployed configs. This is set automatically on Databricks
@@ -189,21 +198,28 @@ else
       # If a master is not specified, use "local[cores, $SPARK_TASK_MAXFAILURES]"
       if [ -z "${PYSP_TEST_spark_master}" ] && [[ "$SPARK_SUBMIT_FLAGS" != *"--master"* ]]; then
         CPU_CORES=`nproc` 
+        # We are limiting the number of tasks in local mode to 4 because it helps to reduce the
+        # total memory usage, especially host memory usage because when copying data to the GPU
+        # buffers as large as batchSizeBytes can be allocated, and the fewer of them we have the better.
         LOCAL_PARALLEL=$(( $CPU_CORES > 4 ? 4 : $CPU_CORES ))
         export PYSP_TEST_spark_master="local[$LOCAL_PARALLEL,$SPARK_TASK_MAXFAILURES]"
       fi
     fi
 
-    # CUDA + RAPIDS has a faiurly large overhead (over 500 MiB) so we keep the memory limit 750 MiB < what we checked for
+    # If you want to change the amount of GPU memory allocated you have to change it here 
+    # and where TEST_PARALLEL is calculated
+    export PYSP_TEST_spark_rapids_memory_gpu_allocSize='1536m'
+
     if ((${#TEST_PARALLEL_OPTS[@]} > 0));
     then
-        export PYSP_TEST_spark_rapids_memory_gpu_allocSize='1536m'
         exec python "${RUN_TESTS_COMMAND[@]}" "${TEST_PARALLEL_OPTS[@]}" "${TEST_COMMON_OPTS[@]}"
     else
+        # We set the GPU memory size to be a constant value even if only running with a parallelism of 1
+        # because it helps us have consistent test runs.
         exec "$SPARK_HOME"/bin/spark-submit --jars "${ALL_JARS// /,}" \
             --driver-java-options "$PYSP_TEST_spark_driver_extraJavaOptions" \
             $SPARK_SUBMIT_FLAGS \
-            --conf 'spark.rapids.memory.gpu.allocSize=1536m' \
+            --conf 'spark.rapids.memory.gpu.allocSize='"$PYSP_TEST_spark_rapids_memory_gpu_allocSize" \
             "${RUN_TESTS_COMMAND[@]}" "${TEST_COMMON_OPTS[@]}"
     fi
 fi

--- a/integration_tests/run_pyspark_from_build.sh
+++ b/integration_tests/run_pyspark_from_build.sh
@@ -63,45 +63,31 @@ else
     fi
     if [[ "${TEST_PARALLEL}" == "" ]];
     then
+        # For integration tests we want to have at least
+        #  - 2 GiB of GPU memory
+        #  - 5 GiB of host memory
+        #  - 1 CPU core
+        # per Spark application. And we want to reserve 2 GiB of GPU memory for CUDA/etc.
+
         # For now just assume that we are going to use the GPU on the
-        # system with the most free memory and then divide it up into chunks.
+        # system with the most free memory.
         # We use free memory to try and avoid issues if the GPU also is working
         # on graphics, which happens some times.
-        # We subtract one for the main controlling process that will still
-        # launch an application.  It will not run thing on the GPU but it needs
-        # to still launch a spark application.
-        TEST_PARALLEL=`nvidia-smi --query-gpu=memory.free --format=csv,noheader | awk '{if (MAX < $1){ MAX = $1}} END {print int(MAX / (2.3 * 1024)) - 1}'`
+        # we also reserve 2 GiB for CUDA/CUDF overhead on the GPU.
+        
+        GPU_MEM_PARALLEL=`nvidia-smi --query-gpu=memory.free --format=csv,noheader | awk '{if (MAX < $1){ MAX = $1}} END {print int((MAX - 2 * 1024)/ ((1.5 * 1024) + 750))}'`
+        CPU_CORES=`nproc` 
+        HOST_MEM_PARALLEL=`cat /proc/meminfo | grep MemAvailable | awk '{print int($2 / (5 * 1024))}'`
+        TMP_PARALLEL=$(( $GPU_MEM_PARALLEL>$CPU_CORES ? $CPU_CORES : $GPU_MEM_PARALLEL ))
+        TMP_PARALLEL=$(( $TMP_PARALLEL>$HOST_MEM_PARALLEL ? $HOST_MEM_PARALLEL : $TMP_PARALLEL ))
+        if [[ $TMP_PARALLEL -gt 1 ]];
+        then
+            TEST_PARALLEL=$(( $TMP_PARALLEL - 1 ))
+        else
+            TEST_PARALLEL=1
+        fi
+
         echo "AUTO DETECTED PARALLELISM OF $TEST_PARALLEL"
-
-        # adjust TEST_PARALLEL according to cpu cores and free memory;
-        # first read from ENV; Note it's hardly to get CPU cores and real free memory in docker.
-        # So in docker Env, Please set TEST_PARALLEL or set PYSP_TEST_CPU_CORES and PYSP_TEST_FREE_MEMORY
-        if [[ "${PYSP_TEST_CPU_CORES}" != "" ]]; then
-          cpu_cores=${PYSP_TEST_CPU_CORES}
-        else
-          # "nproc" get the number of processing units available
-          # "nproc" will get right number in docker created by "docker run --cpuset-cpus" command,
-          # but not work with "--cpus" or "--cpu-period" or "--cpu-quota".
-          cpu_cores=$(nproc)
-        fi
-        if [[ "${PYSP_TEST_FREE_MEMORY}" != "" ]]; then
-          free_mem_mib=${PYSP_TEST_FREE_MEMORY}
-        else
-          free_mem_mib=$(awk '/MemFree/ { printf "%d\n", $2/1024 }' /proc/meminfo)
-        fi
-
-        # reserve one to avoid 100% usages of all CPU cores
-        max_parallel_for_cpu_cores=$((${cpu_cores} - 1))
-        # default heap size for spark is 1G, add reserve 50M for other processes in OS.
-        max_parallel_for_free_memory=$(( (${free_mem_mib} - 50) / 1024 ))
-        if [[ TEST_PARALLEL -gt $max_parallel_for_cpu_cores ]]; then
-          echo "set TEST_PARALLEL from $TEST_PARALLEL to $max_parallel_for_cpu_cores according to cpu cores $cpu_cores"
-          TEST_PARALLEL=${max_parallel_for_cpu_cores}
-        fi
-        if [[ TEST_PARALLEL -gt ${max_parallel_for_free_memory} ]]; then
-          echo "set TEST_PARALLEL from $TEST_PARALLEL to $max_parallel_for_free_memory according to free memory $free_mem_mib mib"
-          TEST_PARALLEL=${max_parallel_for_free_memory}
-        fi
     fi
     if python -c 'import findspark';
     then
@@ -129,9 +115,7 @@ else
         # With xdist 0 and 1 are the same parallelsm but
         # 0 is more effecient
         TEST_PARALLEL_OPTS=()
-        MEMORY_FRACTION='1'
     else
-        MEMORY_FRACTION=`python -c "print(1/($TEST_PARALLEL + 1))"`
         TEST_PARALLEL_OPTS=("-n" "$TEST_PARALLEL")
     fi
     RUN_DIR=${RUN_DIR-"$SCRIPTPATH"/target/run_dir}
@@ -172,9 +156,12 @@ else
     export PYSP_TEST_spark_executor_extraJavaOptions='-ea -Duser.timezone=UTC'
     export PYSP_TEST_spark_ui_showConsoleProgress='false'
     export PYSP_TEST_spark_sql_session_timeZone='UTC'
-    export PYSP_TEST_spark_sql_shuffle_partitions='12'
+    export PYSP_TEST_spark_sql_shuffle_partitions='4'
     # prevent cluster shape to change
     export PYSP_TEST_spark_dynamicAllocation_enabled='false'
+    export PYSP_TEST_spark_rapids_memory_host_spillStorageSize='100m'
+    # Not the default 2G but should be large enough for a single batch for all data
+    export PYSP_TEST_spark_rapids_sql_batchSizeBytes='100m'
 
     # Extract Databricks version from deployed configs. This is set automatically on Databricks
     # notebooks but not when running Spark manually.
@@ -199,23 +186,24 @@ else
     if ((NUM_LOCAL_EXECS > 0)); then
       export PYSP_TEST_spark_master="local-cluster[$NUM_LOCAL_EXECS,$CORES_PER_EXEC,$MB_PER_EXEC]"
     else
-      # If a master is not specified, use "local[*, $SPARK_TASK_MAXFAILURES]"
+      # If a master is not specified, use "local[cores, $SPARK_TASK_MAXFAILURES]"
       if [ -z "${PYSP_TEST_spark_master}" ] && [[ "$SPARK_SUBMIT_FLAGS" != *"--master"* ]]; then
-        export PYSP_TEST_spark_master="local[*,$SPARK_TASK_MAXFAILURES]"
+        CPU_CORES=`nproc` 
+        LOCAL_PARALLEL=$(( $CPU_CORES > 4 ? 4 : $CPU_CORES ))
+        export PYSP_TEST_spark_master="local[$LOCAL_PARALLEL,$SPARK_TASK_MAXFAILURES]"
       fi
     fi
 
+    # CUDA + RAPIDS has a faiurly large overhead (over 500 MiB) so we keep the memory limit 750 MiB < what we checked for
     if ((${#TEST_PARALLEL_OPTS[@]} > 0));
     then
-        export PYSP_TEST_spark_rapids_memory_gpu_allocFraction=$MEMORY_FRACTION
-        export PYSP_TEST_spark_rapids_memory_gpu_maxAllocFraction=$MEMORY_FRACTION
-        # when running tests in parallel, we allocate less than the default minAllocFraction per test
-        # so we need to override this setting here
-        export PYSP_TEST_spark_rapids_memory_gpu_minAllocFraction=0
+        export PYSP_TEST_spark_rapids_memory_gpu_allocSize='1536m'
         exec python "${RUN_TESTS_COMMAND[@]}" "${TEST_PARALLEL_OPTS[@]}" "${TEST_COMMON_OPTS[@]}"
     else
         exec "$SPARK_HOME"/bin/spark-submit --jars "${ALL_JARS// /,}" \
             --driver-java-options "$PYSP_TEST_spark_driver_extraJavaOptions" \
-            $SPARK_SUBMIT_FLAGS "${RUN_TESTS_COMMAND[@]}" "${TEST_COMMON_OPTS[@]}"
+            $SPARK_SUBMIT_FLAGS \
+            --conf 'spark.rapids.memory.gpu.allocSize=1536m' \
+            "${RUN_TESTS_COMMAND[@]}" "${TEST_COMMON_OPTS[@]}"
     fi
 fi

--- a/integration_tests/src/main/python/csv_test.py
+++ b/integration_tests/src/main/python/csv_test.py
@@ -170,7 +170,7 @@ _empty_double_schema = StructType([
 _enable_all_types_conf = {'spark.rapids.sql.csvTimestamps.enabled': 'true',
         'spark.sql.legacy.timeParserPolicy': 'CORRECTED'}
 
-def read_csv_df(data_path, schema, options = {}):
+def read_csv_df(data_path, schema, spark_tmp_table_factory_ignored, options = {}):
     def read_impl(spark):
         reader = spark.read
         if not schema is None:
@@ -180,13 +180,13 @@ def read_csv_df(data_path, schema, options = {}):
         return debug_df(reader.csv(data_path))
     return read_impl
 
-def read_csv_sql(data_path, schema, options = {}):
+def read_csv_sql(data_path, schema, spark_tmp_table_factory, options = {}):
     opts = options
     if not schema is None:
         opts = copy_and_update(options, {'schema': schema})
     def read_impl(spark):
-        spark.sql('DROP TABLE IF EXISTS `TMP_CSV_TABLE`')
-        return spark.catalog.createTable('TMP_CSV_TABLE', source='csv', path=data_path, **opts)
+        tmp_name = spark_tmp_table_factory.get()
+        return spark.catalog.createTable(tmp_name, source='csv', path=data_path, **opts)
     return read_impl
 
 @approximate_float
@@ -246,12 +246,12 @@ def read_csv_sql(data_path, schema, options = {}):
 @pytest.mark.parametrize('read_func', [read_csv_df, read_csv_sql])
 @pytest.mark.parametrize('v1_enabled_list', ["", "csv"])
 @pytest.mark.parametrize('ansi_enabled', ["true", "false"])
-def test_basic_csv_read(std_input_path, name, schema, options, read_func, v1_enabled_list, ansi_enabled):
+def test_basic_csv_read(std_input_path, name, schema, options, read_func, v1_enabled_list, ansi_enabled, spark_tmp_table_factory):
     updated_conf=copy_and_update(_enable_all_types_conf, {
         'spark.sql.sources.useV1SourceList': v1_enabled_list,
         'spark.sql.ansi.enabled': ansi_enabled
     })
-    assert_gpu_and_cpu_are_equal_collect(read_func(std_input_path + '/' + name, schema, options),
+    assert_gpu_and_cpu_are_equal_collect(read_func(std_input_path + '/' + name, schema, spark_tmp_table_factory, options),
             conf=updated_conf)
 
 csv_supported_gens = [
@@ -285,7 +285,7 @@ def test_round_trip(spark_tmp_path, data_gen, v1_enabled_list):
 @allow_non_gpu('org.apache.spark.sql.execution.LeafExecNode')
 @pytest.mark.parametrize('read_func', [read_csv_df, read_csv_sql])
 @pytest.mark.parametrize('disable_conf', ['spark.rapids.sql.format.csv.enabled', 'spark.rapids.sql.format.csv.read.enabled'])
-def test_csv_fallback(spark_tmp_path, read_func, disable_conf):
+def test_csv_fallback(spark_tmp_path, read_func, disable_conf, spark_tmp_table_factory):
     data_gens =[
         StringGen('(\\w| |\t|\ud720){0,10}', nullable=False),
         byte_gen, short_gen, int_gen, long_gen, boolean_gen, date_gen]
@@ -296,7 +296,7 @@ def test_csv_fallback(spark_tmp_path, read_func, disable_conf):
     schema = gen.data_type
     updated_conf = copy_and_update(_enable_all_types_conf, {disable_conf: 'false'})
 
-    reader = read_func(data_path, schema)
+    reader = read_func(data_path, schema, spark_tmp_table_factory)
     with_cpu_session(
             lambda spark : gen_df(spark, gen).write.csv(data_path))
     assert_gpu_fallback_collect(

--- a/integration_tests/src/main/python/json_test.py
+++ b/integration_tests/src/main/python/json_test.py
@@ -72,7 +72,7 @@ _date_schema = StructType([
 _string_schema = StructType([
     StructField('a', StringType())])
 
-def read_json_df(data_path, schema, options = {}):
+def read_json_df(data_path, schema, spark_tmp_table_factory_ignored, options = {}):
     def read_impl(spark):
         reader = spark.read
         if not schema is None:
@@ -82,13 +82,13 @@ def read_json_df(data_path, schema, options = {}):
         return debug_df(reader.json(data_path))
     return read_impl
 
-def read_json_sql(data_path, schema, options = {}):
+def read_json_sql(data_path, schema, spark_tmp_table_factory, options = {}):
     opts = options
     if not schema is None:
         opts = copy_and_update(options, {'schema': schema})
     def read_impl(spark):
-        spark.sql('DROP TABLE IF EXISTS `TMP_json_TABLE`')
-        return spark.catalog.createTable('TMP_json_TABLE', source='json', path=data_path, **opts)
+        tmp_name = spark_tmp_table_factory.get()
+        return spark.catalog.createTable(tmp_name, source='json', path=data_path, **opts)
     return read_impl
 
 @approximate_float
@@ -217,13 +217,14 @@ def test_json_ts_formats_round_trip(spark_tmp_path, date_format, ts_part, v1_ena
 @pytest.mark.parametrize('allow_non_numeric_numbers', ["true", "false"])
 @pytest.mark.parametrize('allow_numeric_leading_zeros', ["true"])
 @pytest.mark.parametrize('ansi_enabled', ["true", "false"])
-def test_basic_json_read(std_input_path, filename, schema, read_func, allow_non_numeric_numbers, allow_numeric_leading_zeros, ansi_enabled):
+def test_basic_json_read(std_input_path, filename, schema, read_func, allow_non_numeric_numbers, allow_numeric_leading_zeros, ansi_enabled, spark_tmp_table_factory):
     updated_conf = copy_and_update(_enable_all_types_conf,
         {'spark.sql.ansi.enabled': ansi_enabled,
          'spark.sql.legacy.timeParserPolicy': 'CORRECTED'})
     assert_gpu_and_cpu_are_equal_collect(
         read_func(std_input_path + '/' + filename,
         schema,
+        spark_tmp_table_factory,
         { "allowNonNumericNumbers": allow_non_numeric_numbers,
           "allowNumericLeadingZeros": allow_numeric_leading_zeros}),
         conf=updated_conf)
@@ -240,12 +241,12 @@ def test_basic_json_read(std_input_path, filename, schema, read_func, allow_non_
     'CORRECTED',
     'EXCEPTION'
 ])
-def test_json_read_valid_dates(std_input_path, filename, schema, read_func, ansi_enabled, time_parser_policy):
+def test_json_read_valid_dates(std_input_path, filename, schema, read_func, ansi_enabled, time_parser_policy, spark_tmp_table_factory):
     updated_conf = copy_and_update(_enable_all_types_conf,
                                    {'spark.sql.ansi.enabled': ansi_enabled,
                                     'spark.sql.legacy.timeParserPolicy': time_parser_policy,
                                     'spark.rapids.sql.incompatibleDateFormats.enabled': True})
-    f = read_func(std_input_path + '/' + filename, schema, {})
+    f = read_func(std_input_path + '/' + filename, schema, spark_tmp_table_factory, {})
     if time_parser_policy == 'LEGACY' and ansi_enabled == 'true':
         assert_gpu_fallback_collect(
             f,
@@ -266,11 +267,11 @@ def test_json_read_valid_dates(std_input_path, filename, schema, read_func, ansi
     'CORRECTED',
     'EXCEPTION'
 ])
-def test_json_read_invalid_dates(std_input_path, filename, schema, read_func, ansi_enabled, time_parser_policy):
+def test_json_read_invalid_dates(std_input_path, filename, schema, read_func, ansi_enabled, time_parser_policy, spark_tmp_table_factory):
     updated_conf = copy_and_update(_enable_all_types_conf,
                                    {'spark.sql.ansi.enabled': ansi_enabled,
                                     'spark.sql.legacy.timeParserPolicy': time_parser_policy })
-    f = read_func(std_input_path + '/' + filename, schema, {})
+    f = read_func(std_input_path + '/' + filename, schema, spark_tmp_table_factory, {})
     if time_parser_policy == 'EXCEPTION':
         assert_gpu_and_cpu_error(
             df_fun=lambda spark: f(spark).collect(),
@@ -288,10 +289,11 @@ def test_json_read_invalid_dates(std_input_path, filename, schema, read_func, an
 @pytest.mark.parametrize('read_func', [read_json_df, read_json_sql])
 @pytest.mark.parametrize('allow_unquoted_chars', ["true"])
 @pytest.mark.parametrize('filename', ['unquotedChars.json'])
-def test_json_unquotedCharacters(std_input_path, filename, schema, read_func, allow_unquoted_chars):
+def test_json_unquotedCharacters(std_input_path, filename, schema, read_func, allow_unquoted_chars, spark_tmp_table_factory):
     assert_gpu_and_cpu_are_equal_collect(
         read_func(std_input_path + '/' + filename,
         schema,
+        spark_tmp_table_factory,
         {"allowUnquotedControlChars": allow_unquoted_chars}),
         conf=_enable_all_types_conf)
 

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala
@@ -176,49 +176,55 @@ object GpuDeviceManager extends Logging {
   private def computeRmmPoolSize(conf: RapidsConf, info: CudaMemInfo): Long = {
     def truncateToAlignment(x: Long): Long = x & ~511L
 
-    val minAllocation = truncateToAlignment((conf.rmmAllocMinFraction * info.total).toLong)
-    val maxAllocation = truncateToAlignment((conf.rmmAllocMaxFraction * info.total).toLong)
-    val reserveAmount = if (GpuShuffleEnv.shouldUseRapidsShuffle(conf)
-        && conf.rmmPool.equalsIgnoreCase("ASYNC")) {
-      // When using the async allocator, UCX calls `cudaMalloc` directly to allocate the
-      // bounce buffers.
-      conf.rmmAllocReserve + conf.shuffleUcxBounceBuffersSize * 2
-    } else {
-      conf.rmmAllocReserve
-    }
-    var poolAllocation = truncateToAlignment(
-      (conf.rmmAllocFraction * (info.free - reserveAmount)).toLong)
-    if (poolAllocation < minAllocation) {
-      throw new IllegalArgumentException(s"The pool allocation of " +
-        s"${toMB(poolAllocation)} MB (calculated from ${RapidsConf.RMM_ALLOC_FRACTION} " +
-        s"(=${conf.rmmAllocFraction}) and ${toMB(info.free)} MB free memory) was less than " +
-        s"the minimum allocation of ${toMB(minAllocation)} (calculated from " +
-        s"${RapidsConf.RMM_ALLOC_MIN_FRACTION} (=${conf.rmmAllocMinFraction}) " +
-        s"and ${toMB(info.total)} MB total memory)")
-    }
-    if (maxAllocation < poolAllocation) {
-      throw new IllegalArgumentException(s"The pool allocation of " +
-        s"${toMB(poolAllocation)} MB (calculated from ${RapidsConf.RMM_ALLOC_FRACTION} " +
-        s"(=${conf.rmmAllocFraction}) and ${toMB(info.free)} MB free memory) was more than " +
-        s"the maximum allocation of ${toMB(maxAllocation)} (calculated from " +
-        s"${RapidsConf.RMM_ALLOC_MAX_FRACTION} (=${conf.rmmAllocMaxFraction}) " +
-        s"and ${toMB(info.total)} MB total memory)")
-    }
-    if (reserveAmount >= maxAllocation) {
-      throw new IllegalArgumentException(s"RMM reserve memory (${toMB(reserveAmount)} MB) " +
-          s"larger than maximum pool size (${toMB(maxAllocation)} MB). Check the settings for " +
-          s"${RapidsConf.RMM_ALLOC_MAX_FRACTION} (=${conf.rmmAllocFraction}) and " +
-          s"${RapidsConf.RMM_ALLOC_RESERVE} (=$reserveAmount)")
-    }
-    val adjustedMaxAllocation = truncateToAlignment(maxAllocation - reserveAmount)
-    if (poolAllocation > adjustedMaxAllocation) {
-      logWarning(s"RMM pool allocation (${toMB(poolAllocation)} MB) does not leave enough free " +
-          s"memory for reserve memory (${toMB(reserveAmount)} MB), lowering the pool size to " +
-          s"${toMB(adjustedMaxAllocation)} MB to accommodate the requested reserve amount.")
-      poolAllocation = adjustedMaxAllocation
-    }
+    // No checks when rmmExactAlloc is given. We are just going to go with the amount requested
+    // with the proper alignment (which is a requirement for some allocators). This is because
+    // it is for testing and we assume that the tests know what they are doing. If the conf becomes
+    // public, then we need to do some more work.
+    conf.rmmExactAlloc.map(truncateToAlignment).getOrElse {
+      val minAllocation = truncateToAlignment((conf.rmmAllocMinFraction * info.total).toLong)
+      val maxAllocation = truncateToAlignment((conf.rmmAllocMaxFraction * info.total).toLong)
+      val reserveAmount = if (GpuShuffleEnv.shouldUseRapidsShuffle(conf)
+          && conf.rmmPool.equalsIgnoreCase("ASYNC")) {
+        // When using the async allocator, UCX calls `cudaMalloc` directly to allocate the
+        // bounce buffers.
+        conf.rmmAllocReserve + conf.shuffleUcxBounceBuffersSize * 2
+      } else {
+        conf.rmmAllocReserve
+      }
+      var poolAllocation = truncateToAlignment(
+        (conf.rmmAllocFraction * (info.free - reserveAmount)).toLong)
+      if (poolAllocation < minAllocation) {
+        throw new IllegalArgumentException(s"The pool allocation of " +
+            s"${toMB(poolAllocation)} MB (calculated from ${RapidsConf.RMM_ALLOC_FRACTION} " +
+            s"(=${conf.rmmAllocFraction}) and ${toMB(info.free)} MB free memory) was less than " +
+            s"the minimum allocation of ${toMB(minAllocation)} (calculated from " +
+            s"${RapidsConf.RMM_ALLOC_MIN_FRACTION} (=${conf.rmmAllocMinFraction}) " +
+            s"and ${toMB(info.total)} MB total memory)")
+      }
+      if (maxAllocation < poolAllocation) {
+        throw new IllegalArgumentException(s"The pool allocation of " +
+            s"${toMB(poolAllocation)} MB (calculated from ${RapidsConf.RMM_ALLOC_FRACTION} " +
+            s"(=${conf.rmmAllocFraction}) and ${toMB(info.free)} MB free memory) was more than " +
+            s"the maximum allocation of ${toMB(maxAllocation)} (calculated from " +
+            s"${RapidsConf.RMM_ALLOC_MAX_FRACTION} (=${conf.rmmAllocMaxFraction}) " +
+            s"and ${toMB(info.total)} MB total memory)")
+      }
+      if (reserveAmount >= maxAllocation) {
+        throw new IllegalArgumentException(s"RMM reserve memory (${toMB(reserveAmount)} MB) " +
+            s"larger than maximum pool size (${toMB(maxAllocation)} MB). Check the settings for " +
+            s"${RapidsConf.RMM_ALLOC_MAX_FRACTION} (=${conf.rmmAllocFraction}) and " +
+            s"${RapidsConf.RMM_ALLOC_RESERVE} (=$reserveAmount)")
+      }
+      val adjustedMaxAllocation = truncateToAlignment(maxAllocation - reserveAmount)
+      if (poolAllocation > adjustedMaxAllocation) {
+        logWarning(s"RMM pool allocation (${toMB(poolAllocation)} MB) does not leave enough free " +
+            s"memory for reserve memory (${toMB(reserveAmount)} MB), lowering the pool size to " +
+            s"${toMB(adjustedMaxAllocation)} MB to accommodate the requested reserve amount.")
+        poolAllocation = adjustedMaxAllocation
+      }
 
-    poolAllocation
+      poolAllocation
+    }
   }
 
   private def initializeRmm(gpuId: Int, rapidsConf: Option[RapidsConf] = None): Unit = {

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala
@@ -340,6 +340,13 @@ object RapidsConf {
     .checkValue(v => v >= 0 && v <= 1, "The fraction value must be in [0, 1].")
     .createWithDefault(1)
 
+  val RMM_EXACT_ALLOC = conf("spark.rapids.memory.gpu.allocSize")
+      .doc("The exact size in byte that RMM should allocate. This is intended to only be " +
+          "used for testing.")
+      .internal() // If this becomes public we need to add in checks for the value when it is used.
+      .bytesConf(ByteUnit.BYTE)
+      .createOptional
+
   val RMM_ALLOC_MAX_FRACTION = conf(RMM_ALLOC_MAX_FRACTION_KEY)
     .doc("The fraction of total GPU memory that limits the maximum size of the RMM pool. " +
         s"The value must be greater than or equal to the setting for $RMM_ALLOC_FRACTION. " +
@@ -1501,6 +1508,8 @@ class RapidsConf(conf: Map[String, String]) extends Logging {
       pool
     }
   }
+
+  lazy val rmmExactAlloc: Option[Long] = get(RMM_EXACT_ALLOC)
 
   lazy val rmmAllocFraction: Double = get(RMM_ALLOC_FRACTION)
 

